Upload A File/Download Upload Library.gcomp/javascript/UploadFileToWebVI.js

```diff
@@ -12,46 +12,96 @@
-
 (function () {
 	// Use strict prevents silent and common JavaScript errors.
 	'use strict';
 
-	window.OpenFileWebVI = function () {
-		var inputButton = document.getElementById(inputElementName);
-		inputButton.click();
-		inputButton.onchange = function () {
-			window.LoadFile();
+	var waitForClickAndRedirectCancellable = function (elementToIntercept, elementToTarget) {
+		var cancel;
+		var occurred = new Promise(function (resolve) {
+			var handler = function () {
+				elementToIntercept.removeEventListener('click', handler);
+				elementToTarget.click();
+				resolve();
+			};
+			cancel = function () {
+				elementToIntercept.removeEventListener('click', handler);
+				resolve();
+			};
+			elementToIntercept.addEventListener('click', handler);
+		});
+		return {
+			occurred,
+			cancel
 		};
 	}
 
-	window.LoadFile = function () {
-		var myFile = document.getElementById(inputElementName).files[0];
-		var reader = new FileReader();
-
-		reader.addEventListener('load', function (e) {
-			fileRead = false;
-			fileText = e.target.result;
+	var waitForChange = function (input) {
+		return new Promise(function (resolve) {
+			var handler = function () {
+				input.removeEventListener('change', handler);
+				resolve();
+			};
+			input.addEventListener('change', handler);
 		});
-
-		reader.readAsBinaryString(myFile);
-		reader.onload = function (e) {
-			fileRead = true;
-		};
 	}
 
-	window.CheckIfDone = function () {
-		return fileRead;
+	var fileSelected = async function (element, input) {
+		var click = waitForClickAndRedirectCancellable(element, input);
+		var change = waitForChange(input); // Note, intentionally avoid await
+
+		// User must attempt to select a file first
+		await click.occurred;
+
+		do {
+			// Create observer for retry click
+			click = waitForClickAndRedirectCancellable(element, input); // Note, intentionally avoid await
+			// Wait for retry click or value change
+			await Promise.race([change, click.occurred]);
+			// Clear pending retry click if the user changed the value
+			click.cancel();
+		} while (input.files.length === 0);
 	}
 
-	window.ReadText = function () {
-		fileRead = false;
-		return fileText;
+	var readFile = async function (file) {
+		var fileUrl = URL.createObjectURL(file);
+		try {
+			var response = await fetch(fileUrl);
+			var text = await response.text();
+			return text;
+		} finally {
+			URL.revokeObjectURL(fileUrl);
+		}
 	}
 
-}());
+	// OpenFileWebVI lets you pass any selector that targets a single element
+	// We find the element, wait for it to be clicked, and then forward that click to a hidden file input to trigger a file dialog
+	// Unfortunately, we cannot tell if the user clicks cancel on the dialog.
+	// To workaround we listen for two different events:
+	// 1. "change" if a user did select a file
+	// 2. another "click" event because the user clicked cancel (can't observe) and then tried to select a file again
+	window.OpenFileWebVI = async function (selector) {
```

Comment:
This is the core of the change. Switched to making the OpenFileWebVI function to being async. WebVIs understand the async keyword / Promises (the async keyword is built on top of Promises) and will wait for the async function to finish before moving on in the block diagram. This allows us to avoid polling loops to check if a file was selected.